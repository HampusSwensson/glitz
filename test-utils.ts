import { posix, relative } from 'path';
import * as ts from 'typescript';

export function typescriptDiagnostics(filename: string) {
  const program = ts.createProgram([filename], {
    target: ts.ScriptTarget.ESNext,
    moduleResolution: ts.ModuleResolutionKind.NodeJs,
    jsx: ts.JsxEmit.Preserve,
    strict: true,
    baseUrl: __dirname,
    paths: {
      '@glitz/*': ['./packages/*/src'],
      '@glitz/type': ['./packages/type'],
    },
    noEmit: true,
  });

  const diagnostics = ts.getPreEmitDiagnostics(program);

  return diagnostics
    .filter(diagnostic => diagnostic.file && diagnostic.start)
    .map(diagnostic => {
      const { line, character } = diagnostic.file!.getLineAndCharacterOfPosition(diagnostic.start!);
      const path = relative(__dirname, diagnostic.file!.fileName);
      return `${posix.join(...path.split('\\'))} ${line}:${character} - ${ts.flattenDiagnosticMessageText(
        diagnostic.messageText,
        '\n',
      )}`;
    });
}
