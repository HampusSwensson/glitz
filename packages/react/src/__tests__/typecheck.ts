import { resolve } from 'path';
import { typescriptDiagnostics } from '../../../../test-utils';

test('typings', () => {
  const errors = typescriptDiagnostics(resolve(__dirname, '__fixtures__/typings.tsx'));
  expect(errors.length).toBe(14);
  for (const error of errors) {
    expect(error).toMatchSnapshot();
  }
});
