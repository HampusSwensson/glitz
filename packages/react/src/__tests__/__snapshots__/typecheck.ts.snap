// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typings 1`] = `
"packages/react/src/__tests__/__fixtures__/typings.tsx 203:1 - Type '{ a: true; b: true; }' is not assignable to type '(IntrinsicAttributes & { a: boolean; b?: undefined; } & { css?: Style | Style[] | StyledDecorator | undefined; } & { children?: ReactNode; }) | (IntrinsicAttributes & ... 2 more ... & { ...; })'.
  Type '{ a: true; b: true; }' is not assignable to type '{ a?: undefined; b: boolean; }'.
    Types of property 'a' are incompatible.
      Type 'true' is not assignable to type 'undefined'."
`;

exports[`typings 2`] = `
"packages/react/src/__tests__/__fixtures__/typings.tsx 205:30 - Type '{ color: \\"\\"; unknownProperty: number; }' is not assignable to type 'Style | Style[] | StyledDecorator | undefined'.
  Object literal may only specify known properties, and 'unknownProperty' does not exist in type 'Style | Style[] | StyledDecorator'."
`;

exports[`typings 3`] = `
"packages/react/src/__tests__/__fixtures__/typings.tsx 207:12 - Type '{ color: number; }' is not assignable to type 'Style | Style[] | StyledDecorator | undefined'.
  Types of property 'color' are incompatible.
    Type 'number' is not assignable to type '\\"transparent\\" | \\"inherit\\" | \\"-moz-initial\\" | \\"initial\\" | \\"revert\\" | \\"unset\\" | \\"aliceblue\\" | \\"antiquewhite\\" | \\"aqua\\" | \\"aquamarine\\" | \\"azure\\" | \\"beige\\" | \\"bisque\\" | \\"black\\" | \\"blanchedalmond\\" | ... 171 more ... | undefined'."
`;

exports[`typings 4`] = `
"packages/react/src/__tests__/__fixtures__/typings.tsx 209:24 - Argument of type '{ color: \\"\\"; unknownProperty: number; }' is not assignable to parameter of type 'Style'.
  Object literal may only specify known properties, and 'unknownProperty' does not exist in type 'Style'."
`;

exports[`typings 5`] = `"packages/react/src/__tests__/__fixtures__/typings.tsx 211:13 - Type 'number' is not assignable to type '\\"transparent\\" | \\"inherit\\" | \\"-moz-initial\\" | \\"initial\\" | \\"revert\\" | \\"unset\\" | \\"aliceblue\\" | \\"antiquewhite\\" | \\"aqua\\" | \\"aquamarine\\" | \\"azure\\" | \\"beige\\" | \\"bisque\\" | \\"black\\" | \\"blanchedalmond\\" | ... 171 more ... | undefined'."`;

exports[`typings 6`] = `
"packages/react/src/__tests__/__fixtures__/typings.tsx 213:20 - No overload matches this call.
  The last overload gave the following error.
    Argument of type '{ color: \\"\\"; unknownProperty: number; }' is not assignable to parameter of type 'Style'.
      Object literal may only specify known properties, and 'unknownProperty' does not exist in type 'Style'."
`;

exports[`typings 7`] = `
"packages/react/src/__tests__/__fixtures__/typings.tsx 215:9 - No overload matches this call.
  The last overload gave the following error.
    Type 'number' is not assignable to type '\\"transparent\\" | \\"inherit\\" | \\"-moz-initial\\" | \\"initial\\" | \\"revert\\" | \\"unset\\" | \\"aliceblue\\" | \\"antiquewhite\\" | \\"aqua\\" | \\"aquamarine\\" | \\"azure\\" | \\"beige\\" | \\"bisque\\" | \\"black\\" | \\"blanchedalmond\\" | ... 171 more ... | undefined'."
`;

exports[`typings 8`] = `
"packages/react/src/__tests__/__fixtures__/typings.tsx 217:43 - Type '{ color: \\"\\"; unknownProperty: number; }' is not assignable to type 'Style | Style[] | StyledDecorator | undefined'.
  Object literal may only specify known properties, and 'unknownProperty' does not exist in type 'Style | Style[] | StyledDecorator'."
`;

exports[`typings 9`] = `
"packages/react/src/__tests__/__fixtures__/typings.tsx 219:25 - Type '{ color: number; }' is not assignable to type 'Style | Style[] | StyledDecorator | undefined'.
  Types of property 'color' are incompatible.
    Type 'number' is not assignable to type '\\"transparent\\" | \\"inherit\\" | \\"-moz-initial\\" | \\"initial\\" | \\"revert\\" | \\"unset\\" | \\"aliceblue\\" | \\"antiquewhite\\" | \\"aqua\\" | \\"aquamarine\\" | \\"azure\\" | \\"beige\\" | \\"bisque\\" | \\"black\\" | \\"blanchedalmond\\" | ... 171 more ... | undefined'."
`;

exports[`typings 10`] = `
"packages/react/src/__tests__/__fixtures__/typings.tsx 221:51 - No overload matches this call.
  The last overload gave the following error.
    Argument of type '{ color: \\"\\"; unknownProperty: number; }' is not assignable to parameter of type 'Style'.
      Object literal may only specify known properties, and 'unknownProperty' does not exist in type 'Style'."
`;

exports[`typings 11`] = `
"packages/react/src/__tests__/__fixtures__/typings.tsx 223:40 - No overload matches this call.
  The last overload gave the following error.
    Type 'number' is not assignable to type '\\"transparent\\" | \\"inherit\\" | \\"-moz-initial\\" | \\"initial\\" | \\"revert\\" | \\"unset\\" | \\"aliceblue\\" | \\"antiquewhite\\" | \\"aqua\\" | \\"aquamarine\\" | \\"azure\\" | \\"beige\\" | \\"bisque\\" | \\"black\\" | \\"blanchedalmond\\" | ... 171 more ... | undefined'."
`;

exports[`typings 12`] = `
"packages/react/src/__tests__/__fixtures__/typings.tsx 225:7 - No overload matches this call.
  The last overload gave the following error.
    Argument of type '() => Promise<unknown>' is not assignable to parameter of type 'FunctionComponent<{}> | StyledElementLike<ComponentType<StyledElementProps>> | ComponentClass<{}, any>'.
      Type '() => Promise<unknown>' is not assignable to type 'FunctionComponent<{}>'.
        Type 'Promise<unknown>' is missing the following properties from type 'ReactElement<any, any>': type, props, key"
`;

exports[`typings 13`] = `
"packages/react/src/__tests__/__fixtures__/typings.tsx 227:7 - No overload matches this call.
  The last overload gave the following error.
    Argument of type '() => Promise<unknown>' is not assignable to parameter of type 'FunctionComponent<{}> | StyledElementLike<ComponentType<StyledElementProps>> | ComponentClass<{}, any>'.
      Type '() => Promise<unknown>' is not assignable to type 'FunctionComponent<{}>'.
        Type 'Promise<unknown>' is not assignable to type 'ReactElement<any, any>'."
`;

exports[`typings 14`] = `
"packages/react/src/__tests__/__fixtures__/typings.tsx 229:1 - 'styled.button' cannot be used as a JSX component.
  Its return type 'StyledComponentWithRef<ButtonHTMLAttributes<HTMLButtonElement>, HTMLButtonElement>' is not a valid JSX element.
    Type 'StyledComponentWithRef<ButtonHTMLAttributes<HTMLButtonElement>, HTMLButtonElement>' is missing the following properties from type 'Element': type, props, key"
`;

exports[`typings 15`] = `"packages/react/src/__tests__/__fixtures__/typings.tsx 232:1 - JSX element type 'NotAStyledComponent' does not have any construct or call signatures."`;

exports[`typings 16`] = `
"packages/react/src/__tests__/__fixtures__/typings.tsx 244:44 - Type '{ css: {}; ref: RefObject<unknown>; }' is not assignable to type 'IntrinsicAttributes & Pick<{ ref?: ((instance: HTMLButtonElement | null) => void) | RefObject<HTMLButtonElement> | null | undefined; }, never> & { ...; } & { ...; }'.
  Property 'ref' does not exist on type 'IntrinsicAttributes & Pick<{ ref?: ((instance: HTMLButtonElement | null) => void) | RefObject<HTMLButtonElement> | null | undefined; }, never> & { ...; } & { ...; }'."
`;
